dnsdist: Add support for early DoH HTTP responses

rgacogne opened the pull request:
### Short description
<!-- Write a small description of what this Pull Request fixes or provides, including the issue #s -->
The `HTTPPathRule` and `HTTPPathRegexRule`, and the `Lua` bindings to access the HTTP informations are only invoked for valid DNS over HTTP queries. This PR adds support for responding to HTTP queries _before_ the DNS payload has been parsed, thus allowing to respond to all HTTP queries.

### Checklist
<!-- please indicate if any of these things are done/included with this Pull Request. Not all boxes need to be checked for the Pull Request to be accepted -->
I have:
- [x] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [x] compiled this code
- [x] tested this code
- [x] included documentation (including possible behaviour changes)
- [x] documented the code
- [x] added or modified regression test(s)
- [ ] added or modified unit test(s)

chbruyand commented on pdns/dnsdistdist/doh.cc:
```diff
@@ -470,6 +574,13 @@ try
 
+  for (const auto& entry : dsc->df->d_responsesMap) {
+    if (entry->matches(path)) {
+      handleResponse(*dsc->df, req, entry->getStatusCode(), entry->getContent(), std::string());
```
Content-type will be set to `application/dns-message` which may not be what we want

chbruyand replied:
`Content-type` header will be set by default depending on the desired status code.
This would be `application/dns-message` in case of a 200  response which may not be what we want.

rgacogne commented:
> I'm just wondering if matching shouldn't be also done on HTTP method ?

It might be a request at some point, yes, but for now I'm trying to keep it simple.

rgacogne commented:
> I'm also wondering if we may want to send the possibly set custom headers in the response ?

I'd really like dnsdist not to become a full-fledged HTTP server :-)

chbruyand commented:
> > I'm also wondering if we may want to send the possibly set custom headers in the response ?
> 
> I'd really like dnsdist not to become a full-fledged HTTP server :-)

I know, this is just that you may expect to get the custom headers you set via `customResponseHeaders` in the response.

rgacogne commented:
> I know, this is just that you may expect to get the custom headers you set via `customResponseHeaders` in the response.

Oh, you are right! I'm not 100% sure it's a good idea to apply these to "early" responses, but I guess it might be something that users will expect..

chbruyand commented:
> > I know, this is just that you may expect to get the custom headers you set via `customResponseHeaders` in the response.
> 
> Oh, you are right! I'm not 100% sure it's a good idea to apply these to "early" responses, but I guess it might be something that users will expect..

Yes, really depends on your custom headers usage. But I think you would expect it in use cases like https://github.com/PowerDNS/pdns/issues/7900

rgacogne commented:
I'm worried about the case where one does not want the `customResponseHeaders` applied here, so I guess we indeed need to accept an optional list of custom headers in `newDOHResponseMapEntry()`, where no list means to apply `customResponseHeaders` and any other value overrides it (so an empty list means no headers).

rgacogne commented:
Fixed the default `application/dns-message` `content-type` header added to early 200 responses. Added support for custom HTTP headers in early responses as well.

rgacogne commented:
Rebased to fix a conflict.
